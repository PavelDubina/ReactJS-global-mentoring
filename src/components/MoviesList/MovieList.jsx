import React, { useState, useEffect, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory, useParams } from 'react-router-dom'
import { MovieCard } from '../MovieCard/MovieCard'
import { Navbar } from '../Navbar/Navbar'
import { EditForm } from '../EditForm/EditForm'
import { DeleteForm } from '../DeleteForm/DeleteForm'
import { ScrollToTop } from '../ScrollToTop/ScrollToTop'
import fetchMovies from '../../redux/actions/fetchMovies'
import { Loading } from '../Loading/Loading'
import styles from './MovieList.scss'

export const MovieList = () => {
  const dispatch = useDispatch()
  const history = useHistory()
  const { query } = useParams()
  const { movies, isLoading, error, sortBy, filter } = useSelector((state) => state.moviesData)
  const [genreValue, setGenreValue] = useState(filter)
  const [sortValue, setSortValue] = useState(sortBy)
  const [{ isOpen, status, movieData }, setMovie] = useState({
    isOpen: false,
    status: 'delete || edit',
    movieData: {},
  })
  const toggleSortValue = (event) => setSortValue(event.target.textContent)
  const handleMenu = (genre) => setGenreValue(genre)
  const onCloseModal = useCallback(() => setMovie((state) => ({ ...state, isOpen: false })), [isOpen])
  const getMovie = (id, status, isOpen) => {
    setMovie({ status, isOpen, movieData: movies.find((movie) => movie.id === id) })
  }
  useEffect(() => {
    dispatch(
      fetchMovies({
        filter: genreValue,
        sortBy: sortValue,
        search: query,
      }),
    )
  }, [genreValue, sortValue, query])
  const handleMovieDetails = (id) => {
    history.push(`/film/${id}`)
    window.scrollTo({ top: 0, behavior: 'smooth' })
  }
  return (
    <div className={styles.container}>
      <div className={styles.section}>
        <Navbar
          toggleSortValue={toggleSortValue}
          handleMenu={handleMenu}
          genreValue={genreValue}
          sortValue={sortValue}
        />
        <div className={styles.result__count}>
          <span className={styles.result__count_number}>{movies.length}</span>
          <span>movies found</span>
        </div>
        <div className={styles.list}>
          {movies.map((movie) => (
            <MovieCard
              poster={movie.poster_path}
              genres={movie.genres}
              title={movie.title}
              date={movie.release_date}
              vote={movie.vote_average}
              overview={movie.overview}
              runtime={movie.runtime}
              id={movie.id}
              key={movie.id}
              getMovie={getMovie}
              handleMovieDetails={handleMovieDetails}
            />
          ))}
          {status === 'Edit' && isOpen && (
            <EditForm
              onClose={onCloseModal}
              genres={movieData.genres}
              title={movieData.title}
              id={movieData.id}
              release_date={movieData.release_date}
              overview={movieData.overview}
              runtime={movieData.runtime}
              poster_path={movieData.poster_path}
              vote_average={movieData.vote_average}
              vote_count={movieData.vote_count}
            />
          )}
          {status === 'Delete' && isOpen && <DeleteForm id={movieData.id} onClose={onCloseModal} />}
        </div>
        {error && <p className={styles.not_found}>Movies not found</p>}
        {isLoading && <Loading />}
      </div>
      <ScrollToTop />
    </div>
  )
}
