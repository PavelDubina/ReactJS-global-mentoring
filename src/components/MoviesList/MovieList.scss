.container {
  flex-grow: 1;
  background-color: #232323;
  padding: 20px 48px 32px;
}

.section {
  max-width: 1280px;
  margin: 0 auto;
}

.list {
  display: grid;
  grid-template-columns: repeat(auto-fit, minmax(260px, 1fr));
  gap: 32px;
}

.result__count {
  padding: 16px 0;
  font-size: 18px;
  color: #ffffffc2;

  &_number {
    font-weight: 700;
    margin: 0 5px 0 0;
  }
}
