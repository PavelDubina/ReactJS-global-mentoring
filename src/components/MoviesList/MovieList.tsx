import React, { useState, useEffect, useCallback } from 'react'
import { useRouter } from 'next/router'
import { useDispatch, useSelector } from 'react-redux'
import { useTranslation } from 'next-i18next'
import { MovieCard } from '../MovieCard/MovieCard'
import { Navbar } from '../Navbar/Navbar'
import { EditForm } from '../EditForm/EditForm'
import { DeleteForm } from '../DeleteForm/DeleteForm'
import { ScrollToTop } from '../ScrollToTop/ScrollToTop'
import fetchMovies from '../../redux/actions/fetchMovies'
import { Loading } from '../Loading/Loading'
import styles from '../../../styles/MovieList.module.scss'
import { TAppState } from '../../redux/reducers/rootReducer'
import { TMovie } from '../../redux/types'

type TUseMovieState = {
  isOpen: boolean
  status: string
  movieData: TMovie
}

export const MovieList: React.FC = () => {
  const { t } = useTranslation()
  const dispatch = useDispatch()
  const router = useRouter()
  const { query, id } = router.query
  const { movies, isLoading, error, sortBy, filter, search } = useSelector((state: TAppState) => state.moviesData)
  const [genreValue, setGenreValue] = useState(filter)
  const [sortValue, setSortValue] = useState(sortBy)
  const [{ isOpen, status, movieData }, setMovie] = useState<TUseMovieState>({
    isOpen: false,
    status: 'delete || edit',
    movieData: {
      title: '',
      tagline: '',
      vote_average: 0,
      vote_count: 0,
      release_date: '',
      poster_path: '',
      overview: '',
      budget: 0,
      revenue: 0,
      runtime: 0,
      genres: [''],
      id: 0,
    },
  })
  const toggleSortValue = (sortValue: string) => setSortValue(sortValue)
  const handleMenu = (genre: string) => setGenreValue(genre)
  const onCloseModal = useCallback(() => setMovie((state) => ({ ...state, isOpen: false })), [isOpen])
  const getMovie = (id: number, status: string, isOpen: boolean) => {
    setMovie({ status, isOpen, movieData: movies.find((movie) => movie.id === id) || movieData })
  }
  useEffect(() => {
    if (router.isReady) {
      dispatch(
        fetchMovies({
          filter: genreValue,
          sortBy: sortValue,
          search: id ? search : (query as string),
        }),
      )
    }
  }, [genreValue, sortValue, query, id])
  const handleMovieDetails = (id: number) => {
    router.push({ pathname: `/film/[id]`, query: { id } }, undefined, { scroll: false, shallow: true })
    window.scrollTo({ top: 0, behavior: 'smooth' })
  }
  return (
    <div className={styles.container}>
      <div className={styles.section}>
        <Navbar
          toggleSortValue={toggleSortValue}
          handleMenu={handleMenu}
          genreValue={genreValue}
          sortValue={sortValue}
        />
        <div className={styles.result__count}>
          <span className={styles.result__count_number}>{movies.length}</span>
          <span>{t('found')}</span>
        </div>
        <div className={styles.list}>
          {movies.map((movie) => (
            <MovieCard
              poster={movie.poster_path}
              genres={movie.genres}
              title={movie.title}
              date={movie.release_date}
              id={movie.id}
              key={movie.id}
              getMovie={getMovie}
              handleMovieDetails={handleMovieDetails}
            />
          ))}
          {status === 'edit' && isOpen && (
            <EditForm
              onClose={onCloseModal}
              genres={movieData.genres}
              title={movieData.title}
              id={movieData.id}
              release_date={movieData.release_date}
              overview={movieData.overview}
              runtime={movieData.runtime}
              poster_path={movieData.poster_path}
              vote_average={movieData.vote_average}
              vote_count={movieData.vote_count}
            />
          )}
          {status === 'delete' && isOpen && <DeleteForm id={movieData.id} onClose={onCloseModal} />}
        </div>
        {error && <p className={styles.not_found}>Movies not found</p>}
        {isLoading && <Loading />}
      </div>
      <ScrollToTop />
    </div>
  )
}
