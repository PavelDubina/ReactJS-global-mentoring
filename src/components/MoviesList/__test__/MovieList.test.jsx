import React from 'react'
import ReactDOM from 'react-dom'
import { useRouter } from 'next/router'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { render, fireEvent } from '@testing-library/react'
import movies from '../../../MockedData/mockMovies.json'
import { MovieList } from '../MovieList'

const mockStore = configureStore([])
jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({ push: jest.fn(), query: { id: '3133369', query: '' } }),
}))
describe('MovieList', () => {
  ReactDOM.createPortal = jest.fn((element) => element)
  const store = mockStore({
    moviesData: { movies, error: false, isLoading: false, sortBy: 'RELEASE DATE', filter: 'all' },
  })
  store.dispatch = jest.fn()
  const getComponent = () =>
    render(
      <Provider store={store}>
        <MovieList />
      </Provider>,
    )
  it('Should renders correctly', () => {
    expect(getComponent().asFragment()).toMatchSnapshot()
  })
  it('Should renders correctly with loading', () => {
    const store = mockStore({
      moviesData: { movies: [], error: false, isLoading: true, sortBy: 'RELEASE DATE', filter: 'all' },
    })
    store.dispatch = jest.fn()
    const getComponent = () =>
      render(
        <Provider store={store}>
          <MovieList />
        </Provider>,
      )
    expect(getComponent().asFragment()).toMatchSnapshot()
  })
  it('Should renders correctly with not movies error', () => {
    const store = mockStore({
      moviesData: { movies: [], error: true, isLoading: false, sortBy: 'RELEASE DATE', filter: 'all' },
    })
    store.dispatch = jest.fn()
    const getComponent = () =>
      render(
        <Provider store={store}>
          <MovieList />
        </Provider>,
      )
    const { getByText } = getComponent()
    expect(getByText('Movies not found')).toBeTruthy()
  })
  it('Should handle movie details and should change pathname', () => {
    window.scrollTo = jest.fn()
    const { getByAltText } = getComponent()
    fireEvent.click(getByAltText('poster'))
    expect(useRouter().push).toBeCalled()
  })
  it('Should call toggleSortValue', () => {
    const { getByTestId } = getComponent()
    const select = getByTestId('select')
    fireEvent.click(select)
    expect(getByTestId('RATING').classList.contains('active')).toBeFalsy()
    fireEvent.click(getByTestId('RATING'))
    expect(getByTestId('RATING').classList.contains('active')).toBeTruthy()
  })
  it('Should call handleMenu', () => {
    const { getByText } = getComponent()
    const comedyNav = getByText('comedy')
    expect(comedyNav.classList.contains('active')).toBeFalsy()
    fireEvent.click(comedyNav)
    expect(comedyNav.classList.contains('active')).toBeTruthy()
  })
  it('Should call onCloseModal and getMovie, open delete modal', () => {
    const { getByText, getByTestId } = getComponent()
    const cardMenu = getByTestId('card-menu')
    fireEvent.click(cardMenu)
    const deleteModalBtn = getByText('Delete')
    fireEvent.click(deleteModalBtn)
    expect(getByText('DELETE MOVIE')).toBeTruthy()
    fireEvent.click(getByText('Ã—'))
  })
  it('Should open edit modal', () => {
    const { getByText, getByTestId } = getComponent()
    const cardMenu = getByTestId('card-menu')
    fireEvent.click(cardMenu)
    const editModalBtn = getByText('Edit')
    fireEvent.click(editModalBtn)
    expect(getByText('EDIT MOVIE')).toBeTruthy()
  })
})
