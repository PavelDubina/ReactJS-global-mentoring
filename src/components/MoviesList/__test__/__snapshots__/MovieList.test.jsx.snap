// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="section"
    >
      <nav
        class="container"
      >
        <div
          class="genres"
        >
          <div
            class="genre active"
          >
            all
          </div>
          <div
            class="genre"
          >
            documentary
          </div>
          <div
            class="genre"
          >
            comedy
          </div>
          <div
            class="genre"
          >
            horror
          </div>
          <div
            class="genre"
          >
            crime
          </div>
        </div>
        <div
          class="sort"
        >
          <span>
            SORT BY
          </span>
          <div
            class="select"
            data-testid="select"
          >
            RELEASE DATE
            <div
              class="dropdown"
              data-testid="dropdown"
            >
              <p
                class="active"
                data-testid="RELEASE DATE"
              >
                RELEASE DATE
              </p>
              <p
                class=""
                data-testid="RATING"
              >
                RATING
              </p>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="result__count"
      >
        <span
          class="result__count_number"
        >
          1
        </span>
        <span>
          movies found
        </span>
      </div>
      <div
        class="list"
      >
        <div
          class="container"
        >
          <div
            class="poster"
          >
            <img
              alt="poster"
              height="16"
              src="https://image.tmdb.org/t/p/w500/5wioUjzwr91fvq7ydunxK2CPSxi.jpg"
              width="9"
            />
            <div
              class="menu"
              data-testid="menu"
            >
              <div
                class="menu"
                data-testid="card-menu"
              >
                <div />
                <div />
                <div />
              </div>
            </div>
          </div>
          <div
            class="movie__data"
          >
            <div
              class="movie__data_info"
            >
              <h2
                class="movie__data_info_title"
              >
                Noah
              </h2>
              <p
                class="movie__data_info_genres"
              >
                Drama & Adventure
              </p>
            </div>
            <div
              class="movie__data_date"
            >
              <span>
                2014
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="scroll_to_top"
    />
  </div>
</DocumentFragment>
`;

exports[`MovieList Should renders correctly with loading 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="section"
    >
      <nav
        class="container"
      >
        <div
          class="genres"
        >
          <div
            class="genre active"
          >
            all
          </div>
          <div
            class="genre"
          >
            documentary
          </div>
          <div
            class="genre"
          >
            comedy
          </div>
          <div
            class="genre"
          >
            horror
          </div>
          <div
            class="genre"
          >
            crime
          </div>
        </div>
        <div
          class="sort"
        >
          <span>
            SORT BY
          </span>
          <div
            class="select"
            data-testid="select"
          >
            RELEASE DATE
            <div
              class="dropdown"
              data-testid="dropdown"
            >
              <p
                class="active"
                data-testid="RELEASE DATE"
              >
                RELEASE DATE
              </p>
              <p
                class=""
                data-testid="RATING"
              >
                RATING
              </p>
            </div>
          </div>
        </div>
      </nav>
      <div
        class="result__count"
      >
        <span
          class="result__count_number"
        >
          0
        </span>
        <span>
          movies found
        </span>
      </div>
      <div
        class="list"
      />
      <div
        class="lds_spinner"
      >
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </div>
    </div>
    <div
      class="scroll_to_top"
    />
  </div>
</DocumentFragment>
`;
